sqlmesh/core/table_diff.py

```diff
@@ -191,23 +180,67 @@ def __init__(
             for col in ensure_list(skip_columns)
         }
 
-        normalize_identifiers(self.on, dialect=self.model_dialect or self.dialect)
-
-        self._source_schema: t.Optional[t.Dict[str, exp.DataType]] = None
-        self._target_schema: t.Optional[t.Dict[str, exp.DataType]] = None
+        self._on = on
         self._row_diff: t.Optional[RowDiff] = None
 
-    @property
+    @cached_property
     def source_schema(self) -> t.Dict[str, exp.DataType]:
-        if self._source_schema is None:
-            self._source_schema = self.adapter.columns(self.source_table)
-        return self._source_schema
+        return self.adapter.columns(self.source_table)
 
-    @property
+    @cached_property
     def target_schema(self) -> t.Dict[str, exp.DataType]:
-        if self._target_schema is None:
-            self._target_schema = self.adapter.columns(self.target_table)
-        return self._target_schema
+        return self.adapter.columns(self.target_table)
+
+    @cached_property
+    def key_columns(self) -> t.Tuple[t.List[exp.Column], t.List[exp.Column], t.List[str]]:
+        dialect = self.model_dialect or self.dialect
+
+        # If the columns to join on are explicitly specified, then just return them
+        if isinstance(self._on, (list, tuple)):
+            identifiers = [normalize_identifiers(c, dialect=dialect) for c in self._on]
+            s_index = [exp.column(c, "s") for c in identifiers]
+            t_index = [exp.column(c, "t") for c in identifiers]
+            return s_index, t_index, [i.name for i in identifiers]
+
+        # Otherwise, we need to parse them out of the supplied "on" condition
+        index_cols = []
+        s_index = []
+        t_index = []
+
+        normalize_identifiers(self._on, dialect=dialect)
+        for col in self._on.find_all(exp.Column):
+            index_cols.append(col.name)
+            if col.table.lower() == "s":
+                s_index.append(col)
+            elif col.table.lower() == "t":
+                t_index.append(col)
+
+        index_cols = list(dict.fromkeys(index_cols))
+        s_index = list(dict.fromkeys(s_index))
+        t_index = list(dict.fromkeys(t_index))
+
+        return s_index, t_index, index_cols
+
+    @property
+    def source_key_expression(self) -> exp.Expression:
+        s_index, _, _ = self.key_columns
+        return self._key_expression(s_index, self.source_schema)
+
+    @property
+    def target_key_expression(self) -> exp.Expression:
+        _, t_index, _ = self.key_columns
+        return self._key_expression(t_index, self.target_schema)
+
+    def _key_expression(
+        self, cols: t.List[exp.Column], schema: t.Dict[str, exp.DataType]
+    ) -> exp.Expression:
+        # if there is a single column, dont do anything fancy to it in order to allow existing indexes to be hit
+        if len(cols) == 1:
+            return exp.to_column(cols[0].name)
```

Comment:
[Nit] do we need to reparse the column here? Can we just pass `cols[0]` to this? (That would copy it, but we can also pass `copy=False` if not needed.)

Reply:
If the column is something like `t.foo`, then we want to strip the `t` off it and just have a column `foo` (since it may have originated from the user's arbitrary `--on` expression).

This allows it to be used within the source query where the `t` reference doesn't exist, as well as being aliased into a join condition

Reply:
Gotcha, thanks for clarifying.